### ISSUE TITLE
PathNotationConverter: Convert Provider<> and FileSystemLocation

### ISSUE THREAD

**Issue by mstrangfeld** [gradle/gradle#8751]
### tl;dr
The [`PathNotationConverter`](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/api/internal/file/copy/PathNotationConverter.java) should also be able to convert [`Provider<>`](https://github.com/gradle/gradle/blob/master/subprojects/core-api/src/main/java/org/gradle/api/provider/Provider.java) and [`FileSystemLocation`](https://github.com/gradle/gradle/blob/master/subprojects/core-api/src/main/java/org/gradle/api/file/FileSystemLocation.java)
### The problem
I would like to create a `CopySpec` to configure a `Copy` task later.
To configure it lazily I would like to use the `DirectoryProperty` to set the `into`.
Example:
```kotlin
// build.gradle.kts
val intoDir = objects.directoryProperty()

intoDir.set(File("test"))

val mySpec = copySpec {
    from(files("from"))
    into(intoDir)
}

val copyTask by tasks.registering(Copy::class) {
    into(buildDir)
    with(mySpec)
}
```
`$ gradle copyTask`

### Expected Behavior

The `CopySpec` uses the `DirectoryProperty` to set it's values and resolves the value later.

### Current Behavior

With the latest Gradle version `Gradle 5.2.1`:
```
$ gradle copyTask
> Task :copyTask FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':copyTask'.
> Cannot convert the provided notation to a String: property(interface org.gradle.api.file.Directory, fixed(class org.gradle.api.internal.file.DefaultFilePropertyFactory$FixedDirectory, /tmp/gradle/testCopyWithProperties/test)).
  The following types/formats are supported:
    - String or CharSequence instances, for example 'some/path'.
    - Boolean values, for example true, Boolean.TRUE.
    - Number values, for example 42, 3.14.
    - A File instance
    - A Closure that returns any supported value.
    - A Callable that returns any supported value.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
1 actionable task: 1 executed

```

### Context

The documentation says: 
#### CopySpec
> The destination is evaluated as per Project.file(Object).
> -- <cite>[CopySpec.into(Object destPath)](https://docs.gradle.org/current/javadoc/org/gradle/api/file/CopySpec.html#into-java.lang.Object-)</cite>
#### Project
> Resolves a file path relative to the project directory of this project. This method converts the supplied path based on its type:
>
> + A CharSequence ...
> + ...
> + A Directory or RegularFile.
> + A Provider of any supported type. The provider's value is resolved recursively.
> -- <cite>[Project.file​(Object path)](https://docs.gradle.org/current/javadoc/org/gradle/api/Project.html#file-java.lang.Object-)</cite>

So I would expect that a `DirectoryProperty` works as well.

The problem seems to be the `PathNotationConverter` which does not support `Provider<>` and `FileSystemLocation` at this moment.
